<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Introduction | Analysis workflow for IMC data</title>
  <meta name="description" content="This bookdown project highlights possible down-stream analyses performed on imaging mass cytometry data." />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Introduction | Analysis workflow for IMC data" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This bookdown project highlights possible down-stream analyses performed on imaging mass cytometry data." />
  <meta name="github-repo" content="BodenmillerGroup/IMCDataAnalysis" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Introduction | Analysis workflow for IMC data" />
  
  <meta name="twitter:description" content="This bookdown project highlights possible down-stream analyses performed on imaging mass cytometry data." />
  

<meta name="author" content="Authors: Nils Eling 1,2,*, Vito Zanotelli 1,2, Michelle Daniel 1,2, Daniel Schulz 1,2, Jonas Windhager 1,2, Lasse Meyer 1,2" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="processing.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Multiplexed imaging data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> IMC Data Analysis Workflow</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#disclaimer"><i class="fa fa-check"></i><b>1.1</b> Disclaimer</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#update-freeze"><i class="fa fa-check"></i><b>1.2</b> Update freeze</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#feedback-and-contributing"><i class="fa fa-check"></i><b>1.3</b> Feedback and contributing</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#maintainer"><i class="fa fa-check"></i><b>1.4</b> Maintainer</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#contributors"><i class="fa fa-check"></i><b>1.5</b> Contributors</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#citation"><i class="fa fa-check"></i><b>1.6</b> Citation</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#changelog"><i class="fa fa-check"></i><b>1.7</b> Changelog</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#technical-details-of-imc"><i class="fa fa-check"></i><b>2.1</b> Technical details of IMC</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="intro.html"><a href="intro.html#metal-conjugated-antobodies-and-staining"><i class="fa fa-check"></i><b>2.1.1</b> Metal-conjugated antobodies and staining</a></li>
<li class="chapter" data-level="2.1.2" data-path="intro.html"><a href="intro.html#data-acquisition"><i class="fa fa-check"></i><b>2.1.2</b> Data acquisition</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#data-format"><i class="fa fa-check"></i><b>2.2</b> IMC data format</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="processing.html"><a href="processing.html"><i class="fa fa-check"></i><b>3</b> Multi-channel image processing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="processing.html"><a href="processing.html#image-pre-processing-imc-specific"><i class="fa fa-check"></i><b>3.1</b> Image pre-processing (IMC specific)</a></li>
<li class="chapter" data-level="3.2" data-path="processing.html"><a href="processing.html#image-segmentation"><i class="fa fa-check"></i><b>3.2</b> Image segmentation</a></li>
<li class="chapter" data-level="3.3" data-path="processing.html"><a href="processing.html#feature-extraction"><i class="fa fa-check"></i><b>3.3</b> Feature extraction</a></li>
<li class="chapter" data-level="3.4" data-path="processing.html"><a href="processing.html#data-export"><i class="fa fa-check"></i><b>3.4</b> Data export</a></li>
<li class="chapter" data-level="3.5" data-path="processing.html"><a href="processing.html#data-import-into-r"><i class="fa fa-check"></i><b>3.5</b> Data import into R</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>4</b> Prerequisites</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prerequisites.html"><a href="prerequisites.html#obtain-the-code"><i class="fa fa-check"></i><b>4.1</b> Obtain the code</a></li>
<li class="chapter" data-level="4.2" data-path="prerequisites.html"><a href="prerequisites.html#software-requirements"><i class="fa fa-check"></i><b>4.2</b> Software requirements</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="prerequisites.html"><a href="prerequisites.html#docker"><i class="fa fa-check"></i><b>4.2.1</b> Using Docker</a></li>
<li class="chapter" data-level="4.2.2" data-path="prerequisites.html"><a href="prerequisites.html#manual-install"><i class="fa fa-check"></i><b>4.2.2</b> Manual installation</a></li>
<li class="chapter" data-level="4.2.3" data-path="prerequisites.html"><a href="prerequisites.html#major-package-versions"><i class="fa fa-check"></i><b>4.2.3</b> Major package versions</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="prerequisites.html"><a href="prerequisites.html#image-processing"><i class="fa fa-check"></i><b>4.3</b> Image processing</a></li>
<li class="chapter" data-level="4.4" data-path="prerequisites.html"><a href="prerequisites.html#download-data"><i class="fa fa-check"></i><b>4.4</b> Download example data</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="prerequisites.html"><a href="prerequisites.html#processed-multiplexed-imaging-data"><i class="fa fa-check"></i><b>4.4.1</b> Processed multiplexed imaging data</a></li>
<li class="chapter" data-level="4.4.2" data-path="prerequisites.html"><a href="prerequisites.html#files-for-spillover-matrix-estimation"><i class="fa fa-check"></i><b>4.4.2</b> Files for spillover matrix estimation</a></li>
<li class="chapter" data-level="4.4.3" data-path="prerequisites.html"><a href="prerequisites.html#gated-cells"><i class="fa fa-check"></i><b>4.4.3</b> Gated cells</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="prerequisites.html"><a href="prerequisites.html#sessionInfo"><i class="fa fa-check"></i><b>4.5</b> Software versions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="read-data.html"><a href="read-data.html"><i class="fa fa-check"></i><b>5</b> Read in the data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="read-data.html"><a href="read-data.html#read-in-single-cell-information"><i class="fa fa-check"></i><b>5.1</b> Read in single-cell information</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="read-data.html"><a href="read-data.html#steinbock-generated-data"><i class="fa fa-check"></i><b>5.1.1</b> steinbock generated data</a></li>
<li class="chapter" data-level="5.1.2" data-path="read-data.html"><a href="read-data.html#imc-segmentation-pipeline-generated-data"><i class="fa fa-check"></i><b>5.1.2</b> IMC Segmentation Pipeline generated data</a></li>
<li class="chapter" data-level="5.1.3" data-path="read-data.html"><a href="read-data.html#reading-custom-files"><i class="fa fa-check"></i><b>5.1.3</b> Reading custom files</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="read-data.html"><a href="read-data.html#cell-processing"><i class="fa fa-check"></i><b>5.2</b> Single-cell processing</a></li>
<li class="chapter" data-level="5.3" data-path="read-data.html"><a href="read-data.html#read-images"><i class="fa fa-check"></i><b>5.3</b> Read in images</a></li>
<li class="chapter" data-level="5.4" data-path="read-data.html"><a href="read-data.html#generate-single-cell-data-from-images"><i class="fa fa-check"></i><b>5.4</b> Generate single-cell data from images</a></li>
<li class="chapter" data-level="5.5" data-path="read-data.html"><a href="read-data.html#accessing-publicly-available-imc-datasets"><i class="fa fa-check"></i><b>5.5</b> Accessing publicly available IMC datasets</a></li>
<li class="chapter" data-level="5.6" data-path="read-data.html"><a href="read-data.html#save-objects"><i class="fa fa-check"></i><b>5.6</b> Save objects</a></li>
<li class="chapter" data-level="5.7" data-path="read-data.html"><a href="read-data.html#session-info"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="spillover-correction.html"><a href="spillover-correction.html"><i class="fa fa-check"></i><b>6</b> Spillover correction</a>
<ul>
<li class="chapter" data-level="6.1" data-path="spillover-correction.html"><a href="spillover-correction.html#generate-the-spillover-matrix"><i class="fa fa-check"></i><b>6.1</b> Generate the spillover matrix</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="spillover-correction.html"><a href="spillover-correction.html#read-in-the-data"><i class="fa fa-check"></i><b>6.1.1</b> Read in the data</a></li>
<li class="chapter" data-level="6.1.2" data-path="spillover-correction.html"><a href="spillover-correction.html#quality-control"><i class="fa fa-check"></i><b>6.1.2</b> Quality control</a></li>
<li class="chapter" data-level="6.1.3" data-path="spillover-correction.html"><a href="spillover-correction.html#pixel_binning"><i class="fa fa-check"></i><b>6.1.3</b> Optional pixel binning</a></li>
<li class="chapter" data-level="6.1.4" data-path="spillover-correction.html"><a href="spillover-correction.html#filtering-incorrectly-assigned-pixels"><i class="fa fa-check"></i><b>6.1.4</b> Filtering incorrectly assigned pixels</a></li>
<li class="chapter" data-level="6.1.5" data-path="spillover-correction.html"><a href="spillover-correction.html#compute-spillover-matrix"><i class="fa fa-check"></i><b>6.1.5</b> Compute spillover matrix</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="spillover-correction.html"><a href="spillover-correction.html#single-cell-data-compensation"><i class="fa fa-check"></i><b>6.2</b> Single-cell data compensation</a></li>
<li class="chapter" data-level="6.3" data-path="spillover-correction.html"><a href="spillover-correction.html#image-compensation"><i class="fa fa-check"></i><b>6.3</b> Image compensation</a></li>
<li class="chapter" data-level="6.4" data-path="spillover-correction.html"><a href="spillover-correction.html#write-out-compensated-images"><i class="fa fa-check"></i><b>6.4</b> Write out compensated images</a></li>
<li class="chapter" data-level="6.5" data-path="spillover-correction.html"><a href="spillover-correction.html#save-objects-1"><i class="fa fa-check"></i><b>6.5</b> Save objects</a></li>
<li class="chapter" data-level="6.6" data-path="spillover-correction.html"><a href="spillover-correction.html#session-info-1"><i class="fa fa-check"></i><b>6.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html"><i class="fa fa-check"></i><b>7</b> Image and cell-level quality control</a>
<ul>
<li class="chapter" data-level="7.1" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#read-in-the-data-1"><i class="fa fa-check"></i><b>7.1</b> Read in the data</a></li>
<li class="chapter" data-level="7.2" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#seg-quality"><i class="fa fa-check"></i><b>7.2</b> Segmentation quality control</a></li>
<li class="chapter" data-level="7.3" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#image-quality"><i class="fa fa-check"></i><b>7.3</b> Image-level quality control</a></li>
<li class="chapter" data-level="7.4" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#cell-quality"><i class="fa fa-check"></i><b>7.4</b> Cell-level quality control</a></li>
<li class="chapter" data-level="7.5" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#save-objects-2"><i class="fa fa-check"></i><b>7.5</b> Save objects</a></li>
<li class="chapter" data-level="7.6" data-path="image-and-cell-level-quality-control.html"><a href="image-and-cell-level-quality-control.html#session-info-2"><i class="fa fa-check"></i><b>7.6</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>8</b> Batch effect correction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="batch-effects.html"><a href="batch-effects.html#fastmnn-correction"><i class="fa fa-check"></i><b>8.1</b> fastMNN correction</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="batch-effects.html"><a href="batch-effects.html#perform-sample-correction"><i class="fa fa-check"></i><b>8.1.1</b> Perform sample correction</a></li>
<li class="chapter" data-level="8.1.2" data-path="batch-effects.html"><a href="batch-effects.html#quality-control-of-correction-results"><i class="fa fa-check"></i><b>8.1.2</b> Quality control of correction results</a></li>
<li class="chapter" data-level="8.1.3" data-path="batch-effects.html"><a href="batch-effects.html#visualization"><i class="fa fa-check"></i><b>8.1.3</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="batch-effects.html"><a href="batch-effects.html#harmony-correction"><i class="fa fa-check"></i><b>8.2</b> harmony correction</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="batch-effects.html"><a href="batch-effects.html#visualization-1"><i class="fa fa-check"></i><b>8.2.1</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="batch-effects.html"><a href="batch-effects.html#seurat-correction"><i class="fa fa-check"></i><b>8.3</b> Seurat correction</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="batch-effects.html"><a href="batch-effects.html#visualization-2"><i class="fa fa-check"></i><b>8.3.1</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="batch-effects.html"><a href="batch-effects.html#save-objects-3"><i class="fa fa-check"></i><b>8.4</b> Save objects</a></li>
<li class="chapter" data-level="8.5" data-path="batch-effects.html"><a href="batch-effects.html#session-info-3"><i class="fa fa-check"></i><b>8.5</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html"><i class="fa fa-check"></i><b>9</b> Cell phenotyping</a>
<ul>
<li class="chapter" data-level="9.1" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#load-data"><i class="fa fa-check"></i><b>9.1</b> Load data</a></li>
<li class="chapter" data-level="9.2" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#clustering"><i class="fa fa-check"></i><b>9.2</b> Clustering approaches</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#rphenograph"><i class="fa fa-check"></i><b>9.2.1</b> Rphenograph</a></li>
<li class="chapter" data-level="9.2.2" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#snn-graph"><i class="fa fa-check"></i><b>9.2.2</b> Shared nearest neighbour graph</a></li>
<li class="chapter" data-level="9.2.3" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#self-organizing-maps"><i class="fa fa-check"></i><b>9.2.3</b> Self organizing maps</a></li>
<li class="chapter" data-level="9.2.4" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#compare-between-clustering-approaches"><i class="fa fa-check"></i><b>9.2.4</b> Compare between clustering approaches</a></li>
<li class="chapter" data-level="9.2.5" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#clustering-notes"><i class="fa fa-check"></i><b>9.2.5</b> Further clustering notes</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#classification"><i class="fa fa-check"></i><b>9.3</b> Classification approach</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#manual-labeling-of-cells"><i class="fa fa-check"></i><b>9.3.1</b> Manual labeling of cells</a></li>
<li class="chapter" data-level="9.3.2" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#define-color-vectors"><i class="fa fa-check"></i><b>9.3.2</b> Define color vectors</a></li>
<li class="chapter" data-level="9.3.3" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#read-in-and-consolidate-labeled-data"><i class="fa fa-check"></i><b>9.3.3</b> Read in and consolidate labeled data</a></li>
<li class="chapter" data-level="9.3.4" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#train-classifier"><i class="fa fa-check"></i><b>9.3.4</b> Train classifier</a></li>
<li class="chapter" data-level="9.3.5" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#classifier-performance"><i class="fa fa-check"></i><b>9.3.5</b> Classifier performance</a></li>
<li class="chapter" data-level="9.3.6" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#classification-of-new-data"><i class="fa fa-check"></i><b>9.3.6</b> Classification of new data</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="cell-phenotyping.html"><a href="cell-phenotyping.html#session-info-4"><i class="fa fa-check"></i><b>9.4</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html"><i class="fa fa-check"></i><b>10</b> Single cell visualization</a>
<ul>
<li class="chapter" data-level="10.1" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#load-data-1"><i class="fa fa-check"></i><b>10.1</b> Load data</a></li>
<li class="chapter" data-level="10.2" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#cell-type-level"><i class="fa fa-check"></i><b>10.2</b> Cell-type level</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#dimensionality-reduction-visualization"><i class="fa fa-check"></i><b>10.2.1</b> Dimensionality reduction visualization</a></li>
<li class="chapter" data-level="10.2.2" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#heatmap-visualization"><i class="fa fa-check"></i><b>10.2.2</b> Heatmap visualization</a></li>
<li class="chapter" data-level="10.2.3" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#violin-plot-visualization"><i class="fa fa-check"></i><b>10.2.3</b> Violin plot visualization</a></li>
<li class="chapter" data-level="10.2.4" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#scatter-plot-visualization"><i class="fa fa-check"></i><b>10.2.4</b> Scatter plot visualization</a></li>
<li class="chapter" data-level="10.2.5" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#barplot-visualization"><i class="fa fa-check"></i><b>10.2.5</b> Barplot visualization</a></li>
<li class="chapter" data-level="10.2.6" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#catalyst-based-visualization"><i class="fa fa-check"></i><b>10.2.6</b> CATALYST-based visualization</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#sample-level"><i class="fa fa-check"></i><b>10.3</b> Sample-level</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#dimensionality-reduction-visualization-1"><i class="fa fa-check"></i><b>10.3.1</b> Dimensionality reduction visualization</a></li>
<li class="chapter" data-level="10.3.2" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#heatmap-visualization-1"><i class="fa fa-check"></i><b>10.3.2</b> Heatmap visualization</a></li>
<li class="chapter" data-level="10.3.3" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#barplot-visualization-1"><i class="fa fa-check"></i><b>10.3.3</b> Barplot visualization</a></li>
<li class="chapter" data-level="10.3.4" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#catalyst-based-visualization-1"><i class="fa fa-check"></i><b>10.3.4</b> CATALYST-based visualization</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#rich-example"><i class="fa fa-check"></i><b>10.4</b> Further examples</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#publication-ready-complexheatmap"><i class="fa fa-check"></i><b>10.4.1</b> Publication-ready ComplexHeatmap</a></li>
<li class="chapter" data-level="10.4.2" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#interactive-visualization"><i class="fa fa-check"></i><b>10.4.2</b> Interactive visualization</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="single-cell-visualization.html"><a href="single-cell-visualization.html#session-info-5"><i class="fa fa-check"></i><b>10.5</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="image-visualization.html"><a href="image-visualization.html"><i class="fa fa-check"></i><b>11</b> Image visualization</a>
<ul>
<li class="chapter" data-level="11.1" data-path="image-visualization.html"><a href="image-visualization.html#pixel-visualization"><i class="fa fa-check"></i><b>11.1</b> Pixel visualization</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="image-visualization.html"><a href="image-visualization.html#adjusting-colors"><i class="fa fa-check"></i><b>11.1.1</b> Adjusting colors</a></li>
<li class="chapter" data-level="11.1.2" data-path="image-visualization.html"><a href="image-visualization.html#image-normalization"><i class="fa fa-check"></i><b>11.1.2</b> Image normalization</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="image-visualization.html"><a href="image-visualization.html#mask-visualization"><i class="fa fa-check"></i><b>11.2</b> Cell visualization</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="image-visualization.html"><a href="image-visualization.html#visualzing-metadata"><i class="fa fa-check"></i><b>11.2.1</b> Visualzing metadata</a></li>
<li class="chapter" data-level="11.2.2" data-path="image-visualization.html"><a href="image-visualization.html#visualizating-expression"><i class="fa fa-check"></i><b>11.2.2</b> Visualizating expression</a></li>
<li class="chapter" data-level="11.2.3" data-path="image-visualization.html"><a href="image-visualization.html#outline-cells"><i class="fa fa-check"></i><b>11.2.3</b> Outlining cells on images</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="image-visualization.html"><a href="image-visualization.html#adjusting-plot-annotations"><i class="fa fa-check"></i><b>11.3</b> Adjusting plot annotations</a></li>
<li class="chapter" data-level="11.4" data-path="image-visualization.html"><a href="image-visualization.html#displaying-individual-images"><i class="fa fa-check"></i><b>11.4</b> Displaying individual images</a></li>
<li class="chapter" data-level="11.5" data-path="image-visualization.html"><a href="image-visualization.html#saving-and-returning-images"><i class="fa fa-check"></i><b>11.5</b> Saving and returning images</a></li>
<li class="chapter" data-level="11.6" data-path="image-visualization.html"><a href="image-visualization.html#cytoviewer"><i class="fa fa-check"></i><b>11.6</b> Interactive image visualization</a></li>
<li class="chapter" data-level="11.7" data-path="image-visualization.html"><a href="image-visualization.html#session-info-6"><i class="fa fa-check"></i><b>11.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html"><i class="fa fa-check"></i><b>12</b> Performing spatial analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#spatial-interaction-graphs"><i class="fa fa-check"></i><b>12.1</b> Spatial interaction graphs</a></li>
<li class="chapter" data-level="12.2" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#spatial-viz"><i class="fa fa-check"></i><b>12.2</b> Spatial visualization</a></li>
<li class="chapter" data-level="12.3" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#spatial-community-analysis"><i class="fa fa-check"></i><b>12.3</b> Spatial community analysis</a></li>
<li class="chapter" data-level="12.4" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#cellular-neighborhood-analysis"><i class="fa fa-check"></i><b>12.4</b> Cellular neighborhood analysis</a></li>
<li class="chapter" data-level="12.5" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#spatial-context-analysis"><i class="fa fa-check"></i><b>12.5</b> Spatial context analysis</a></li>
<li class="chapter" data-level="12.6" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#patch-detection"><i class="fa fa-check"></i><b>12.6</b> Patch detection</a></li>
<li class="chapter" data-level="12.7" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#interaction-analysis"><i class="fa fa-check"></i><b>12.7</b> Interaction analysis</a></li>
<li class="chapter" data-level="12.8" data-path="performing-spatial-analysis.html"><a href="performing-spatial-analysis.html#session-info-7"><i class="fa fa-check"></i><b>12.8</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis workflow for IMC data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="intro" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">2</span> Introduction<a href="intro.html#intro" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Highly multiplexed imaging (HMI) enables the simultaneous detection of dozens of
biological molecules (e.g., proteins, transcripts; also referred to as
“markers”) in tissues. Recently established multiplexed tissue imaging
technologies rely on cyclic staining with fluorescently-tagged antibodies
<span class="citation">(<a href="#ref-Lin2018">Lin et al. 2018</a>; <a href="#ref-Gut2018">Gut, Herrmann, and Pelkmans 2018</a>)</span>, or the use of oligonucleotide-tagged <span class="citation">(<a href="#ref-Goltsev2018">Goltsev et al. 2018</a>; <a href="#ref-Saka2019">Saka et al. 2019</a>)</span> or metal-tagged <span class="citation">(<a href="#ref-Giesen2014">Giesen et al. 2014</a>; <a href="#ref-Angelo2014">Angelo et al. 2014</a>)</span> antibodies, among others.
The key strength of these technologies is that they allow in-depth analysis of
single cells within their spatial tissue context. As a result, these methods
have enabled analysis of the spatial architecture of the tumor microenvironment
<span class="citation">(<a href="#ref-Lin2018">Lin et al. 2018</a>; <a href="#ref-Jackson2020">Jackson et al. 2020</a>; <a href="#ref-Ali2020">Ali et al. 2020</a>; <a href="#ref-Schurch2020">Schürch et al. 2020</a>)</span>, determination of nucleic acid
and protein abundances for assessment of spatial co-localization of cell types
and chemokines <span class="citation">(<a href="#ref-Hoch2022">Hoch et al. 2022</a>)</span> and spatial niches of virus infected cells <span class="citation">(<a href="#ref-Jiang2022">Jiang et al. 2022</a>)</span>,
and characterization of pathological features during COVID-19 infection
<span class="citation">(<a href="#ref-Rendeiro2021">Rendeiro et al. 2021</a>; <a href="#ref-Mitamura2021">Mitamura et al. 2021</a>)</span>, Type 1 diabetes progression <span class="citation">(<a href="#ref-Damond2019">Damond et al. 2019</a>)</span> and
autoimmune disease <span class="citation">(<a href="#ref-Ferrian2021">Ferrian et al. 2021</a>)</span>.</p>
<p>Imaging mass cytometry (IMC) utilizes metal-tagged antibodies to detect over 40
proteins and other metal-tagged molecules in biological samples. IMC can be used
to perform highly multiplexed imaging and is particularly suited to profiling
selected areas of tissues across many samples.</p>
<p><img src="img/IMC_workflow.png" alt="IMC_workflow" />
<em>Overview of imaging mass cytometry data acquisition. Taken from <span class="citation">(<a href="#ref-Giesen2014">Giesen et al. 2014</a>)</span></em></p>
<p>IMC has first been published in 2014 <span class="citation">(<a href="#ref-Giesen2014">Giesen et al. 2014</a>)</span> and has been commercialized by
Standard BioTools<sup><font size="1">TM</font></sup> to be distributed as the Hyperion Imaging
System<sup><font size="1">TM</font></sup> (documentation is available
<a href="https://www.fluidigm.com/products-services/instruments/hyperion">here</a>).
Similar to other HMI technologies such as MIBI <span class="citation">(<a href="#ref-Angelo2014">Angelo et al. 2014</a>)</span>, CyCIF <span class="citation">(<a href="#ref-Lin2018">Lin et al. 2018</a>)</span>,
4i <span class="citation">(<a href="#ref-Gut2018">Gut, Herrmann, and Pelkmans 2018</a>)</span>, CODEX <span class="citation">(<a href="#ref-Goltsev2018">Goltsev et al. 2018</a>)</span> and SABER <span class="citation">(<a href="#ref-Saka2019">Saka et al. 2019</a>)</span>, IMC captures the spatial
expression of multiple proteins in parallel. With a nominal 1 μm resolution,
IMC is able to detect cytoplasmic and nuclear localization of proteins. The
current ablation frequency of IMC is 200Hz, meaning that a 1 mm<span class="math inline">\(^2\)</span> area
can be imaged within about 2 hours.</p>
<div id="technical-details-of-imc" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Technical details of IMC<a href="intro.html#technical-details-of-imc" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Technical aspects of how data acquisition works can be found in the original
publication <span class="citation">(<a href="#ref-Giesen2014">Giesen et al. 2014</a>)</span>. Briefly, antibodies to detect targets in biological
material are labeled with heavy metals (e.g., lanthanides) that do not occur in
biological systems and thus can be used upon binding to their target as a
readout similar to fluorophores in fluorescence microscopy. Thin sections of the
biological sample on a glass slide are stained with an antibody cocktail.
Stained microscopy slides are mounted on a precise motor-driven stage inside the
ablation chamber of the IMC instrument. A high-energy UV laser is focused on the
tissue, and each individual laser shot ablates tissue from an area of roughly 1
μm<span class="math inline">\(^2\)</span>. The energy of the laser is absorbed by the tissue resulting
in vaporization followed by condensation of the ablated material. The ablated
material from each laser shot is transported in the gas phase into the plasma of
the mass cytometer, where first atomization of the particles and then ionization
of the atoms occurs. The ion cloud is then transferred into a vacuum, and all
ions below a mass of 80 m/z are filtered using a quadrupole mass filter. The
remaining ions (mostly those used to tag antibodies) are analyzed in a
time-of-flight mass spectrometer to ultimately obtain an accumulated mass
spectrum from all ions that correspond to a single laser shot. One can regard
this spectrum as the information underlying a 1 μm<span class="math inline">\(^2\)</span> pixel. With
repetitive laser shots (e.g., at 200 Hz) and a simultaneous lateral sample
movement, a tissue can be ablated pixel by pixel. Ultimately an image is
reconstructed from each pixel mass spectrum.</p>
<p>In principle, IMC can be applied to the same type of samples as conventional
fluorescence microscopy. The largest distinction from fluorescence microscopy is
that for IMC, primary-labeled antibodies are commonly used, whereas in
fluorescence microscopy secondary antibodies carrying fluorophores are widely
applied. Additionally, for IMC, samples are dried before acquisition and can be
stored for years. Formalin-fixed and paraffin-embedded (FFPE) samples are widely
used for IMC. The FFPE blocks are cut to 2-5 μm thick sections and are
stained, dried, and analyzed with IMC.</p>
<div id="metal-conjugated-antobodies-and-staining" class="section level3 hasAnchor" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> Metal-conjugated antobodies and staining<a href="intro.html#metal-conjugated-antobodies-and-staining" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metal-labeled antibodies are used to stain molecules in tissues enabling to
delineate tissue structures, cells, and subcellular structures. Metal-conjugated
antibodies can either be purchased directly from Standard BioTools<sup><font size="1">TM</font></sup> (<a href="https://store.fluidigm.com/Cytometry/ConsumablesandReagentsCytometry/MaxparAntibodies?cclcl=en_US">MaxPar IMC Antibodies</a>),
or antibodies can be purchased and labeled individually (<a href="https://store.fluidigm.com/Cytometry/ConsumablesandReagentsCytometry/MaxparAntibodyLabelingKits?cclcl=en_US">MaxPar Antibody
Labeling</a>).
Antibody labeling using the MaxPar kits is performed via TCEP antibody reduction
followed by crosslinking with sulfhydryl-reactive maleimide-bearing metal
polymers. For each antibody it is essential to validate its functionality,
specificity and optimize its usage to provide optimal signal to noise. To
facilitate antibody handling, a database is highly useful.
<a href="https://github.com/BodenmillerGroup/airlab-web">Airlab</a> is such a platform; it
allows antibody lot tracking, validation data uploads, and panel generation for
subsequent upload to the IMC acquisition software from Standard BioTools<sup><font size="1">TM</font></sup></p>
<p>Depending on the sample type, different staining protocols can be used.
Generally, once antibodies of choice have been conjugated to a metal tag,
titration experiments are performed to identify the optimal staining
concentration. For FFPE samples, different staining protocols have been
described, and different antibodies show variable staining with different
protocols. Protocols such as the one provided by Standard BioTools<sup><font size="1">TM</font></sup> or the one describe by
<span class="citation">(<a href="#ref-Ijsselsteijn2019">Ijsselsteijn et al. 2019</a>)</span> are recommended. Briefly, for FFPE tissues, a dewaxing
step is performed to remove the paraffin used to embed the material, followed by
a graded re-hydration of the samples. Thereafter, heat-induced epitope retrieval
(HIER), a step aiming at the reversal of formalin-based fixation, is used to
unmask epitopes within tissues and make them accessible to antibodies. Epitope
unmasking is generally performed in either basic, EDTA-based buffers (pH 9.2) or
acidic, citrate-based buffers (pH 6). Next, a buffer containing bovine serum
albumin (BSA) is used to block non-specific binding. This buffer is also used to
dilute antibody stocks for the actual antibody staining. Staining time and
temperature may vary and optimization must be performed to ensure that each
single antibody performs well. However, overnight staining at 4°C or 3-5
hours at room temperature seem to be suitable in many cases.</p>
<p>Following antibody incubation, unbound antibodies are washed away and a
counterstain comparable to DAPI is applied to enable the identification of
nuclei. The <a href="https://store.fluidigm.com/Cytometry/ConsumablesandReagentsCytometry/MassCytometryReagents/Cell-ID%E2%84%A2%20Intercalator-Ir%E2%80%94125%20%C2%B5M">Iridium intercalator</a>
from Standard BioTools<sup><font size="1">TM</font></sup> is a reagent of choice and applied in a brief 5 minute staining.
Finally, the samples are washed again and then dried under an airflow. Once
dried, the samples are ready for analysis using IMC and are
usually stable for a long period of time (at least one year).</p>
</div>
<div id="data-acquisition" class="section level3 hasAnchor" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> Data acquisition<a href="intro.html#data-acquisition" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Data is acquired using the CyTOF software from Standard BioTools<sup><font size="1">TM</font></sup> (see manuals
<a href="https://go.fluidigm.com/hyperion-support-documents">here</a>).</p>
<p>The regions of interest are selected by providing coordinates for ablation. To
determine the region to be imaged, so called “panoramas” can be generated. These
are stitched images of single fields of views of about 200 μm in diameter.
Panoramas provide an optical overview of the tissue with a resolution similar to
10x in microscopy and are intended to help with the selection of regions of
interest for ablation. The tissue should be centered on the glass side, since
the imaging mass cytometer cannot access roughly 5 mm from each of the slide
edges. Currently, the instruments can process one slide at a time and usually one MCD
file per sample slide is generated.</p>
<p>Many regions of interest can be defined on a single slide and acquisition
parameters such as channels to acquire, acquisition speed (100 Hz or 200 Hz),
ablation energy, and other parameters are user-defined. It is recommended that
all isotope channels are recorded. This will result in larger raw data files but valuable information such as
potential contamination of the argon gas (e.g., Xenon) or of the samples (e.g.,
lead, barium) is stored.</p>
<p>To process a large number of slides or to select regions on whole-slide samples,
panoramas may not provide sufficient information. If this is the case,
multi-color immunofluorescence of the same slide prior to staining with
metal-labeled antibodies may be performed. To allow for region selection based
on immunofluorescence images and to align those images with a panorama of the
same or consecutive sections of the sample, we developed
<a href="https://github.com/BodenmillerGroup/napping">napping</a>.</p>
<p>Acquisition time is directly proportional to the total size of ablation, and run
times for samples of large area or for large sample numbers can roughly be calculated by
dividing the ablation area in square micrometer by the ablation speed (e.g.,
200Hz). In addition to the proprietary MCD file format, TXT files can also
be generated for each region of interest. This is recommended as a back-up
option in case of errors that may corrupt MCD files but not TXT files.</p>
</div>
</div>
<div id="data-format" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> IMC data format<a href="intro.html#data-format" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Upon completion of the acquisition an MCD file of variable size is generated. A
single MCD file can hold raw acquisition data for multiple regions of interest,
optical images providing a slide level overview of the sample (“panoramas”), and
detailed metadata about the experiment. Additionally, for each acquisition a
TXT file is generated which holds the same pixel information as the matched
acquisition in the MCD file.</p>
<p>The Hyperion Imaging System<sup><font size="1">TM</font></sup> produces files in the following folder structure:</p>
<pre><code>.
+-- {XYZ}_ROI_001_1.txt
+-- {XYZ}_ROI_002_2.txt
+-- {XYZ}_ROI_003_3.txt
+-- {XYZ}.mcd</code></pre>
<p>Here, <code>{XYZ}</code> defines the filename, <code>ROI_001</code>, <code>ROI_002</code>, <code>ROI_003</code> are
user-defined names (descriptions) for the selected regions of interest (ROI),
and <code>1</code>, <code>2</code>, <code>3</code> indicate the unique acquisition identifiers. The ROI
description entry can be specified in the Standard BioTools software when
selecting ROIs. The MCD file contains the raw imaging data and the full metadata
of all acquired ROIs, while each TXT file contains data of a single ROI without
metadata. To follow a consistent naming scheme and to bundle all metadata, we
recommend to zip the folder. Each ZIP file should only contain data from a
single MCD file, and the name of the ZIP file should match the name of the MCD
file.</p>
<p>We refer to this data as raw data and the further
processing of this data is described in Section <a href="processing.html#processing">3</a>.</p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Ali2020" class="csl-entry">
Ali, Raza, Hartland W. Jackson, Vito R. T. Zanotelli, Esther Danenberg, Jana R. Fischer, Helen Bardwell, Elena Provenzanoa, et al. 2020. <span>“Imaging Mass Cytometry and Multiplatform Genomics Define the Phenogenomic Landscape of Breast Cancer.”</span> <em>Nature Cancer</em> 1: 163–75.
</div>
<div id="ref-Angelo2014" class="csl-entry">
Angelo, Michael, Sean C. Bendall, Rachel Finck, Matthew B. Hale, Chuck Hitzman, Alexander D. Borowsky, Richard M. Levenson, et al. 2014. <span>“Multiplexed Ion Beam Imaging of Human Breast Tumors.”</span> <em>Nature Medicine</em> 20 (4): 436–42.
</div>
<div id="ref-Damond2019" class="csl-entry">
Damond, Nicolas, Stefanie Engler, Vito R. T. Zanotelli, Denis Schapiro, Clive H. Wasserfall, Irina Kusmartseva, Harry S. Nick, et al. 2019. <span>“A Map of Human Type 1 Diabetes Progression by Imaging Mass Cytometry.”</span> <em>Cell Metabolism</em> 29: 755–768.e5.
</div>
<div id="ref-Ferrian2021" class="csl-entry">
Ferrian, Selena, Candace C. Liu, Erin F. McCaffrey, Rashmi Kumar, Theodore S. Nowicki, David W. Dawson, Alex Baranski, et al. 2021. <span>“Multiplexed Imaging Reveals an <span>IFN</span>-<span class="math inline">\(\gamma\)</span>-Driven Inflammatory State in Nivolumab-Associated Gastritis.”</span> <em>Cell Reports Medicine</em> 2: 100419.
</div>
<div id="ref-Giesen2014" class="csl-entry">
Giesen, Charlotte, Hao A. O. Wang, Denis Schapiro, Nevena Zivanovic, Andrea Jacobs, Bodo Hattendorf, Peter J. Schüffler, et al. 2014. <span>“Highly Multiplexed Imaging of Tumor Tissues with Subcellular Resolution by Mass Cytometry.”</span> <em>Nature <span>M</span>ethods</em> 11 (4): 417–22.
</div>
<div id="ref-Goltsev2018" class="csl-entry">
Goltsev, Yury, Nikolay Samusik, Julia Kennedy-Darling, Salil Bhate, Matthew Hale, Gustavo Vazquez, Sarah Black, and Garry P. Nolan. 2018. <span>“Deep Profiling of Mouse Splenic Architecture with CODEX Multiplexed Imaging.”</span> <em>Cell</em> 174: 968–81.
</div>
<div id="ref-Gut2018" class="csl-entry">
Gut, Gabriele, Markus D Herrmann, and Lucas Pelkmans. 2018. <span>“Multiplexed Protein Maps Link Subcellular Organization to Cellular States.”</span> <em>Science</em> 361: 1–13.
</div>
<div id="ref-Hoch2022" class="csl-entry">
Hoch, Tobias, Daniel Schulz, Nils Eling, Julia Martínez Gómez, Mitchell P. Levesque, and Bernd Bodenmiller. 2022. <span>“Multiplexed Imaging Mass Cytometry of the Chemokine Milieus in Melanoma Characterizes Features of the Response to Immunotherapy.”</span> <em>Science Immunology</em> 7 (70): eabk1692.
</div>
<div id="ref-Ijsselsteijn2019" class="csl-entry">
Ijsselsteijn, Marieke E., Ruud van der Breggen, Arantza F. Sarasqueta, Frits Koning, and Noel F. C. C. de Miranda. 2019. <span>“A 40-Marker Panel for High Dimensional Characterization of Cancer Immune Microenvironments by Imaging Mass Cytometry.”</span> <em>Frontiers in Immunology</em> 10.
</div>
<div id="ref-Jackson2020" class="csl-entry">
Jackson, Hartland W., Jana R. Fischer, Vito R. T. Zanotelli, H. Raza Ali, Robert Mechera, Savas D. Soysal, Holger Moch, et al. 2020. <span>“The Single-Cell Pathology Landscape of Breast Cancer.”</span> <em>Nature</em> 578: 615–20.
</div>
<div id="ref-Jiang2022" class="csl-entry">
Jiang, Sizun, Chi Ngai Chan, Xavier Rovira-Clavé, Han Chen, Yunhao Bai, Bokai Zhu, Erin McCaffrey, et al. 2022. <span>“Combined Protein and Nucleic Acid Imaging Reveals Virus-Dependent b Cell and Macrophage Immunosuppression of Tissue Microenvironments.”</span> <em>Immunity</em> 55: 1118–1134.e8.
</div>
<div id="ref-Lin2018" class="csl-entry">
Lin, Jia-Ren, Benjamin Izar, Shu Wang, Clarence Yapp, Shaolin Mei, Parin M. Shah, Sandro Santagata, and Peter K. Sorger. 2018. <span>“Highly Multiplexed Immunofluorescence Imaging of Human Tissues and Tumors Using t-CyCIF and Conventional Optical Microscopes.”</span> <em>eLife</em> 7: 1–46.
</div>
<div id="ref-Mitamura2021" class="csl-entry">
Mitamura, Yasutaka, Daniel Schulz, Saskia Oro, Nick Li, Isabel Kolm, Claudia Lang, Reihane Ziadlou, et al. 2021. <span>“Cutaneous and Systemic Hyperinflammation Drives Maculopapular Drug Exanthema in Severely Ill <span>COVID</span>-19 Patients.”</span> <em>Allergy</em> 77: 595–608.
</div>
<div id="ref-Rendeiro2021" class="csl-entry">
Rendeiro, André F., Hiranmayi Ravichandran, Yaron Bram, Vasuretha Chandar, Junbum Kim, Cem Meydan, Jiwoon Park, et al. 2021. <span>“The Spatial Landscape of Lung Pathology During <span>COVID</span>-19 Progression.”</span> <em>Nature</em> 593: 564–69.
</div>
<div id="ref-Saka2019" class="csl-entry">
Saka, Sinem K., Yu Wang, Jocelyn Y. Kishi, Allen Zhu, Yitian Zeng, Wenxin Xie, Koray Kirli, et al. 2019. <span>“Immuno-SABER Enables Highly Multiplexed and Amplified Protein Imaging in Tissues.”</span> <em>Nature Biotechnology</em> 37: 1080–90.
</div>
<div id="ref-Schurch2020" class="csl-entry">
Schürch, Christian M, Salil S Bhate, Graham L Barlow, Darci J Phillips, Luca Noti, Inti Zlobec, Pauline Chu, et al. 2020. <span>“Coordinated Cellular Neighborhoods Orchestrate Antitumoral Immunity at the Colorectal Cancer Invasive Front.”</span> <em>Cell</em> 182: 1341–59.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="processing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/BodenmillerGroup/IMCDataAnalysis/blob/master/01-intro.Rmd",
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
